src/main/java/bio/terra/stairway/FlightDao.java

```diff
@@ -423,6 +434,66 @@ private void completeInner(FlightContext flightContext) throws SQLException {
     }
   }
 
+  /**
+   * Fatally fail the flight id.
+   *
+   * <p>This should only be used if there are problems creating the FlightContext such that {@link
+   * #complete(FlightContext)} cannot be used.
+   *
+   * <p>This does not do a hook state transition as we do not have a FlightContext.
+   *
+   * @param flightId flight to fail fatally.
+   * @param fatalCause the reason to fail the flight fatally. May be null.
+   */
+  void fatalFail(String flightId, Exception fatalCause)
+      throws DatabaseOperationException, InterruptedException {
+    String serializedException;
+    try {
+      serializedException = exceptionSerializer.serialize(fatalCause);
+    } catch (Exception e) {
+      // We really do not want to throw while we're trying to fatally fail the flight.
+      logger.error("Unable to serialize fatalFail exception.", e);
+      serializedException = "Cause unknown. Unable to serialize fatalFail exception.";
+    }
+    String finalSerializedException = serializedException;
+    DbRetry.retryVoid("flight.fatalFail", () -> fatalFailInner(flightId, finalSerializedException));
+  }
+
+  private void fatalFailInner(String flightId, String serializedException) throws SQLException {
+    // Make this idempotent by not updating if the status is already FATAL.
+    final String sqlUpdateFlight =
+        "UPDATE "
+            + FLIGHT_TABLE
+            + " SET completed_time = CURRENT_TIMESTAMP,"
+            + " status = 'FATAL',"
+            + " serialized_exception = :serializedException,"
+            + " stairway_id = NULL"
+            + " WHERE flightid = :flightId AND status != 'FATAL'";
+
+    // The delete is harmless if it has been done before. We just won't find anything.
+    final String sqlDeleteFlightLog =
+        "DELETE FROM " + FLIGHT_LOG_TABLE + " WHERE flightid = :flightId";
+
+    try (Connection connection = dataSource.getConnection();
+        NamedParameterPreparedStatement statement =
+            new NamedParameterPreparedStatement(connection, sqlUpdateFlight);
+        NamedParameterPreparedStatement deleteStatement =
+            new NamedParameterPreparedStatement(connection, sqlDeleteFlightLog)) {
+
+      startTransaction(connection);
+      statement.setString("serializedException", serializedException);
+      statement.setString("flightId", flightId);
+      statement.getPreparedStatement().executeUpdate();
+
+      if (!keepFlightLog) {
```

Comment:
do we not normally keep the flight logs? I certainly see plenty of them in the database.

Reply:
following what's done in `completeInner`. This is a parameter on Stairway.Builder that can be modified. I won't comment on what's the right default.